hazelcast/src/main/java/com/hazelcast/spi/impl/waitnotifyservice/impl/WaitNotifyServiceImpl.java

```diff
@@ -308,16 +310,35 @@ private boolean doRun() throws Exception {
             }
 
             for (Queue<WaitingOperation> q : mapWaitingOps.values()) {
-                for (WaitingOperation waitingOp : q) {
+                while (it.hasNext()) {
                     if (Thread.interrupted()) {
                         return true;
                     }
+                    WaitingOperation waitingOp = it.next();
                     if (waitingOp.isValid() && waitingOp.needsInvalidation()) {
-                        invalidate(waitingOp);
+                        Address targetAddress = getTargetAddress(waitingOp);
+                        if (!waitingOp.validatesTarget() || nodeEngine.getThisAddress().equals(targetAddress)) {
+                            invalidate(waitingOp);    
+                        } else {
+                            logger.warning("Removing operation " + waitingOp +
+                                " from queue because the local node is not the target and returning exception to invoker");
+                            WrongTargetException wte = new WrongTargetException(nodeEngine.getThisAddress(), targetAddress,
+                                waitingOp.getPartitionId(), waitingOp.getReplicaIndex(),
+                                waitingOp.getClass().getName(), waitingOp.getServiceName());
+                            OperationResponseHandler responseHandler = waitingOp.getOperation().getOperationResponseHandler();
+                            responseHandler.sendResponse(waitingOp.getOperation(), wte);
+                            it.remove();
+                        }
                     }
                 }
             }
             return false;
         }
     }
+
+    private Address getTargetAddress(WaitingOperation waitingOp) {
+        int partitionId = waitingOp.getPartitionId();
+        InternalPartition internalPartition = nodeEngine.getPartitionService().getPartition(partitionId);
+        return internalPartition.getReplicaAddress(waitingOp.getReplicaIndex());
```

Comment:
That's the same logic that throws the WrongTargetException in OperationRunnerImpl:

{quote}
protected void ensureNoPartitionProblems(Operation op) {
        int partitionId = op.getPartitionId();

        if (partitionId < 0) {
            return;
        }

        if (partitionId != getPartitionId()) {
            throw new IllegalStateException("wrong partition, expected: " + getPartitionId() + " but found:" + partitionId);
        }

        if (internalPartition == null) {
            internalPartition = nodeEngine.getPartitionService().getPartition(partitionId);
        }

        if (retryDuringMigration(op) && internalPartition.isMigrating()) {
            throw new PartitionMigratingException(node.getThisAddress(), partitionId,
                    op.getClass().getName(), op.getServiceName());
        }

        Address owner = internalPartition.getReplicaAddress(op.getReplicaIndex());
        if (op.validatesTarget() && !node.getThisAddress().equals(owner)) {
            throw new WrongTargetException(node.getThisAddress(), owner, partitionId, op.getReplicaIndex(),
                    op.getClass().getName(), op.getServiceName());
        }
    }
{quote}